speech/infinite_streaming_from_file/infinite_audio.go

```diff
@@ -0,0 +1,236 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command infinite_audio pipes the stdin audio data to
+// Google Speech API and outputs the transcript.
+
+// Before executing this program, the sampleRate parameter needs to be adjusted
+// to match your audio file, 16000 is used in this program as an example
+
+// To run this program, you could use the following command:
+// cat [your_audio_filename].wav | go run infinite_audio.go
+
+package main
+
+// [START speech_transcribe_streaming]
+import (
+	"context"
+	"io"
+	"log"
+	"math"
+	"os"
+	"sync"
+	"time"
+
+	speech "cloud.google.com/go/speech/apiv1"
+	color "github.com/fatih/color"
+	speechpb "google.golang.org/genproto/googleapis/cloud/speech/v1"
+)
+
+const sampleRate int = 16000                                      //Audio sample rate hertz
+const inputChannel int = 1                                        //Number of audio input channel
+const outputChannel int = 0                                       //Number of audio output channel
+const bytesPerSample int = 2                                      //Number of bytes each sample consists of
+const bytesPerSecond = sampleRate * inputChannel * bytesPerSample //Number of bytes each second audio consists of
+const streamTimeLimit = 290 * time.Second                         //Streaming API Limit( ≈ 5 mins)
+const sampleDuration int64 = 1000                                 //The duration of each sample is 1000 ms
+
+func main() {
```

Comment:
Can we split this into more than one function? It's a bit difficult for me to follow all of the indentation and everything.

Reply:
I try to split into different functions but it is rather difficult to implement the correct logic because some functions need to modify the parameter inside other functions mutually to determine skipping especially when they are running at the same time. So to keep it simple I just combine them into one function and write comment for each sentence of code. Actually there is only one function to implement the main logic and one concurrent function to implement the bi-directional communication. All the other parts of code are auxiliary.